text/0017-message/0017-message.md

```diff
@@ -0,0 +1,166 @@
+|  [Indexation Payload](#indexation-payload)  |     2        |
+
+### Indexation payload
+
+To make the Payload concept clear we will define the `indexation payload`. As the name suggests it allows to add an index to the encapsulating message, as well as some arbitrary data. Nodes will expose an API, that will enable to query messages by the index.
+Adding those capabilities may open nodes to DOS attack vectors:
+1. Proliferation of index keys that may blow up the node's DB
+2. Proliferation of messages associated with the same index
+
+Node implementations may provide weak guarantees regarding the completion of indexes to address the above scenarios. 
+
+Besides the index, the payload will also have a data field.
+  A message that has been attached to the tangle and approved by a milestone has useful properties: You can verify that the content of the data did not change, and you can ascertain the approximate time it was published by checking the approving milestone. If the payload will be incorporated under
+  the `signed transaction payload`, the content will be signed as well.
+
+
+The structure of the payload is simple:
+
+| Name             | Type          | Description               |
+| --------         | -----------   | -----------               |
+| Payload Type     | uint32        | Must be set to **2**      |
+| Index            | string     | The index key of the message |
+| Data             | ByteArray     | Data we are attaching    |
+
+Note that `index` must be at least 1 byte long for the payload to be valid. The `data` may have a length of 0.
+
+### Serialization Example
+
+Below is a serialized valid message with the indexation payload. The index is the "SPAM" ASCII string and the message is the "Hello Iota"
+ASCII string. The [Message PoW Hash](https://github.com/Wollac/protocol-rfcs/blob/message-pow/text/0024-message-pow/0024-message-pow.md) would have
+  10 trailing zeroes for the given nonce in this example. Bytes are expressed as hexadecimal numbers.
+
+[Version] **`01`** [Parent 1] `F532A53545103276B46876C473846D98648EE418468BCE76DF4868648DD73E5D` [Parent 2] `78D546B46AEC4557872139A48F66BC567687E8413578A14323548732358914A2` [Payload Length]
+*`1100000000000000000000000000000000000000000000000000000000000000`*[Payload Type]**`02`**[Index] *`04`*`5350414D` [Data]*`0A`*`48656C6C6F20496F7461`[Nonce]`5d38333333333333`
+
+# Rationale and alternatives
+
+Instead of creating a layered approach, we could have simply created a flat `transaction message`, that is tailored for mutating the ledger state, and try to fit all the use cases there. For example with the unsigned data use-case, we could have filled some section of the transaction with the data. Then via a flag in the transaction we could have instructed to not pass this transaction to the service that attempts to mutate the ledger state.
```

Comment:
```suggestion
Instead of creating a layered approach, we could have simply created a flat `transaction message` that is tailored for mutating the ledger state, and try to fit all the use cases there. For example, with the unsigned data use-case, we could have filled some section of the transaction with the data. Then via a flag in the transaction, we could have instructed to not pass this transaction to the service that attempts to mutate the ledger state.
```